cg/clients/freshdesk/freshdesk_client.py

```diff
@@ -11,28 +18,51 @@
 class FreshdeskClient:
     """Client for communicating with the freshdesk REST API."""
 
-    def __init__(self):
-        self.headers = None
-        self.api_key = None
-        self.url = None
-
-    def init_app(self, url: str, api_key: str):
-        """Set up the client."""
-        self.url = url
+    def __init__(self, base_url: str, api_key: str):
+        self.base_url = base_url
         self.api_key = api_key
-        self.headers = {"Content-Type": "application/json"}
+        self.session = self._configure_session()
+
+    def __enter__(self) -> "FreshdeskClient":
+        return self
 
-    @property
-    def auth_header(self):
-        return self.api_key, "X"
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        self.session.close()
```

Comment:
What is a request session?

Reply:
From https://requests.readthedocs.io/en/latest/user/advanced/#session-objects:

> The Session object allows you to persist certain parameters across requests. It also persists cookies across all requests made from the Session instance, and will use urllib3’s [connection pooling](https://urllib3.readthedocs.io/en/latest/reference/urllib3.connectionpool.html). So if you’re making several requests to the same host, the underlying TCP connection will be reused, which can result in a significant performance increase (see [HTTP persistent connection](https://en.wikipedia.org/wiki/HTTP_persistent_connection)).

Basically you don't have to give parameters to every request, and it also reuses the TCP connection. But the main reason is for configuring retry strategies which seems to be applied on a session rather than individual requests

Reply:
I'm not 100 % sure why it is beneficial to use a context manager like this. Does it not make more sense to instantiate the client once centrally and keep the client lifetime for the lifetime of the app instead of tying it to a `with FreshdeskClient as client` block?

Reply:
We should use the session, I'm just not sure whether we want to tie it to the lifetime of the app or a context manager block

Reply:
After reading up a bit I think I'm convinced. Could you have a look at [f441155](https://github.com/Clinical-Genomics/cg/pull/3328/commits/f441155bc691ac4d12fabbdb15924ac0943a463e)?

Reply:
I'd instantiate the session in the constructor and keep it there to get more value from the pooled TCP connections.

Reply:
And it seems like the package should handle all stuff related to keeping the connection alive automatically https://requests.readthedocs.io/en/latest/user/advanced/#keep-alive

Reply:
Decided to keep one session for the lifetime of the app